### PR TITLE
Added cmdlets to get Pester failures

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
Get-PSPesterFailure will parse the Pester log from start-pspester and extract the failures as objects
Format-PSPesterFailure will format the failures similar to output from Pester (you can use directly without Get-PSPesterFailure)

### CODE REVIEW

File: build.psm1 [hunk 1]
```diff
@@ -627,6 +627,87 @@ function Publish-PSTestTools {
...
+
+        [Parameter(ParameterSetName="object",ValueFromPipeline=$true)]
+        [PSCustomObject]$PesterFailure  
+    )
+
+    begin
+    {
+        if ($PSBoundParameters.Count -eq 0 -or $PSBoundParameters.ContainsKey("NUnitLog"))
+        {
+            $PesterFailure = Get-PSPesterFailures -NUnitLog $NUnitLog
```
Comment (iSazonov) [inline build.psm1]: Get-PSPesterFailure**s** should be replaced by Get-PSPesterFailure

Comment (SteveL-MSFT) [inline build.psm1]: good catch

**Comment by iSazonov** [comment]
LGFM.
(Maybe add in docs how to use this?)

**Comment by SteveL-MSFT** [comment]
@iSazonov good suggestion, I'll add something to the existing testing guidelines

File: docs/testing-guidelines/testing-guidelines.md [hunk 2]
```diff
@@ -106,7 +106,7 @@ As part of your PR, you must ensure all existing tests pass.
 If you see any failures from `Start-PSPester` (summary reported at end), then you must investigate them and fix your PR so those tests pass or fix the test if the test case is wrong.
 Two helper functions are part of the build.psm1 module to help with that:
 * `Get-PSPesterFailure` will parse the NUnit test result log and return PowerShell objects for each failure so you can do additional filtering, sorting, grouping, etc...
-* `Format-PSPesterFailure` will call `Get-PSPesterFailure` if no parameters are provided and show just the failures at the console simliar to what Pester displays
+* `Format-PSPesterFailure` will call `Get-PSPesterFailure` if no parameters are provided and show just the failures at the console similar to what Pester displays
```
Comment (iSazonov) [inline docs/testing-guidelines/testing-guidelines.md]: Maybe add call samples as above for `Start-PSPester`?

File: build.psm1 [hunk 3]
```diff
@@ -627,6 +627,87 @@ function Publish-PSTestTools {
...
+    $outputxml = [xml](get-content $NUnitLog)
+    $failureSuites = SelectNodes -xml $outputxml -xpath "/test-results/test-suite/results/test-suite[@result='Failure']" 
+
+    foreach ($failureSuite in $failureSuites)
+    {
+        $failureTests = SelectNodes -xml $failureSuite -xpath "results/test-case[@result='Failure']"
+        foreach ($failureTest in $failureTests)
+        {
+            $totalFailures++
+            $failure = New-Object -TypeName PSCustomObject -Property @{SuiteName=$failureSuite.Name;TestDescription=$failureTest.Description;
```
Comment (TravisEz13) [inline build.psm1]: Perhaps the properties could be more readable

**Review by TravisEz13 — APPROVED** [review]

File: build.psm1 (hunk 3 above)
Comment (SteveL-MSFT) [inline build.psm1]: @TravisEz13 do you have a specific suggestion as it seems readable to me?

File: docs/testing-guidelines/testing-guidelines.md (hunk 2 above)
Comment (SteveL-MSFT) [inline docs/testing-guidelines/testing-guidelines.md]: added

**Comment by iSazonov** [comment]
In future, it would be good to use it to reduce logs on appveyor and travis-ci to only display failed tests.

**Comment by kilasuit** [comment]
Only query that I would has is should these functions be added to the PowerShell repo or would it make more sense that they be added directly into Pester?

My rational behind adding to Pester is that I can see use cases for them being used outside of just testing the PowerShell Code base so to me it would make more sense for these to be added to Pester.

File: build.psm1 (hunk 3 above)
Comment (TravisEz13) [inline build.psm1]: Guidance developed in the DSC project is to have each name-value pair on a new line

Comment (SteveL-MSFT) [inline build.psm1]: got it, will change

File: docs/testing-guidelines/testing-guidelines.md [hunk 4]
```diff
@@ -84,18 +84,29 @@ When running tests in this way, be sure that you have started PowerShell with `-
 environment is not the default or has any customization.
 
 For example, to run all the Pester tests for CI (assuming you are at the root of the PowerShell repo):
-```
+```PowerShell
```
Comment (SteveL-MSFT) [inline docs/testing-guidelines/testing-guidelines.md]: I'll make this change since I'm already updating this PR

**Comment by iSazonov** [comment]
@kilasuit Your proposal looks very good. I think it would be better to create a Issue in Pester repo and don't block this PR.

**Comment by kilasuit** [comment]
@iSazonov - As much as I agree with your suggestion, I do feel that the onus to do so would lie with @SteveL-MSFT considering that this is his PR & his code etc etc

**Comment by SteveL-MSFT** [comment]
@kilasuit I created https://github.com/pester/Pester/issues/645 as a feature request to Pester.  I think this could be solved in a number of ways and this PR was specifically to solve a problem I encountered where we have a large number of tests for PowerShell (and growing!) and needed an easy way to just see the failures so I can investigate before pushing my changes

**Comment by nohwnd** [comment]
Is this the best way to do it? 

I think you are tapping too late in the pipeline. The pipeline is something like this (pseudo code): 
`Invoke-Pester -PassThrough | Out-NUnitXml | Store-Xml | Parse-XunitReport | Filter-Failures | Print-Fails` 

You are at the end, looking at the nUnit report in XML. Instead you can tap into the pipeline right after Invoke-Pester. Get the result object PSObjects (`-PassThru`), filter to failures with Where, and then either use your print functions, or  you can reuse the internal Pester functions.

```powershell
#store the results in a variable
$results = invoke-pester -passthru

#get first successful result (you'd do -ne "Success")
$result = $results.TestResult | Where {$_.Result -eq "Passed"} | Select -first 1 

#get reference to internal Pester print function (see output.ps1 in Pester, for other output functions)
 $writePesterResult = &(Get-Module Pester) {Get-Command Write-PesterResult}

#print the result 
&$writePesterResult $result
#outputs:  [+] adds positive numbers 44ms
```

Or just wait for us to put it directly in Pester. 😃 I just wanted to highlight what is in my opinion a better way to approach this problem.

**Comment by SteveL-MSFT** [comment]
@nohwnd I think you're right, we can close this PR and wait for pester/Pester#645

**Pull request closed without merging**